giskard/llm/talk/tools/predict.py

```diff
@@ -0,0 +1,153 @@
+        """
+        feature_json_type = get_feature_json_type(self._dataset)
+
+        return {
+            "type": "function",
+            "function": {
+                "name": self.name,
+                "description": self.description,
+                "parameters": {
+                    "type": "object",
+                    "properties": {
+                        "features_dict": {
+                            "type": "object",
+                            "properties": {
+                                feature: {"type": dtype} for feature, dtype in list(feature_json_type.items())
+                            },
+                        }
+                    },
+                    "required": ["features_dict"],
+                },
+            },
+        }
+
+    def _get_input_from_dataset(self, row_filter: dict) -> pd.DataFrame:
+        """Get input from dataset.
+
+        Filter rows from the dataset, using the `row_filter`.
+
+        Parameters
+        ----------
+        row_filter : dict
+            The dictionary with features and related values to filter the dataset.
+
+        Returns
+        -------
+        pd.DataFrame
+            The DataFrame with filtered rows.
+        """
+        threshold = 85
+        filtered_df = self._dataset.df.copy()
+
+        for col_name, col_value in list(row_filter.items()):
+            if filtered_df[col_name].dtype == "object":
+                filtered_df = filtered_df[
+                    filtered_df[col_name].apply(lambda x: fuzz.ratio(x.lower(), col_value.lower()) >= threshold)
+                ]
+                if not len(filtered_df):
+                    break
+            else:
+                filtered_df = filtered_df[filtered_df[col_name] == col_value]
+
+        return filtered_df
+
+    def _get_input_from_user(self, feature_values: dict) -> pd.DataFrame:
+        """Form input based on the user's query.
+
+        Take input feature values from user query. For the missed values, take the feature mean from the Dataset.
+
+        Parameters
+        ----------
+        feature_values : dict
+            The dictionary with features and related values extracted from the query.
+
+        Returns
+        -------
+        pd.DataFrame
+            The DataFrame with the formed input.
+        """
+        # Prepare background sample.
+        from giskard.models.model_explanation import _get_background_example
+
+        background_df = self._model.prepare_dataframe(
+            self._dataset.df, self._dataset.column_dtypes, self._dataset.target
+        )
+        background_sample = _get_background_example(background_df, self._dataset.column_types)
+
+        # Fill background sample with known values.
+        for col_name, col_value in list(feature_values.items()):
+            background_sample.loc[0, col_name] = col_value
+
+        return background_sample
+
+    def _prepare_input(self, feature_values: dict) -> Dataset:
```

Comment:
so here feature values is both a filter and some values ?

Reply:
yes. First, we try to select data using feature values. If there are no rows with such feature values combination, we first calculate mean of each feature of the dataset, and then replace some of them with `feature_values`.